### DISCUSSION TITLE
Demoting Safari to "best effort" support

### DISCUSSION THREAD

**Discussion by sgravrock** [jasmine/jasmine#2050] [General]
Currently, all Safari versions that still receive security updates are treated as fully supported. That's not sustainable, and my intention is to demote **all** versions of Safari to "best effort" support in near-future versions of jasmine-core and jasmine-browser-runner. That will put Safari on the same footing as versions of Node or Firefox ESR that are past end of life.

## What does "best effort" mean?

I'll make reasonable efforts to keep Jasmine working on all supported Safari versions, including running the automated tests on them in CI. However:

* New bugs that only affect Safari might not be treated as release blockers, particularly if they only affect specific Safari versions that I don't have access to.
* I'm more likely to rely on users to contribute Safari-specific bugfixes and enhancements.
* Support for any Safari version may be dropped at any time if it becomes impractical.

## Why?

Mostly because of the cost of keeping enough sufficiently new Macs around. Also partly because Apple doesn't actually publish end of life dates for Safari versions, and estimating support status is difficult.

Safari is uniquely difficult to support. Among Jasmine's supported environments, Safari is:

* The only one that doesn't publish end of life dates
* The only non-evergreen environment that requires a separate OS install for each version
* The only one that can't generally be run in a VM (this may be changing, but it's not a practical option today)
* The only one that can't run on 10+ year old hardware
* The only one that only runs on a single OS

## Why now?

My dedicated Safari test computer has aged out of usefulness, and Jasmine's annual budget of $0 isn't enough to pay for a replacement. That leaves me with local access to only whatever version of Safari is currently on my primary computer. That version potentially changes every time I install OS updates, making it hard to predict what version I'll have access to.

## Why not drop Safari entirely?

Developers who test their work on Safari are doing a good thing both for Safari users and for the web as a whole, and I want to support that.

Running on a broader range of browsers, especially ones that aren't Chrome-based, is good for Jasmine too.

## What's the impact going to be? What should Safari users do?

Hopefully none. Safari-specific bugs are uncommon, and I can do a lot with one local copy of Safari plus Saucelabs. But there's a possibility that new jasmine-core or jasmine-browser-runner versions might ship with known bugs affecting Safari, and support for Safari versions might be dropped sooner than would otherwise have been the case.

If you rely on running Jasmine in Safari, you might want to read the release notes of all jasmine-core and jasmine-browser-runner releases, not just majors. That's especially true if you use older versions of Safari. Please also consider keeping an eye on GitHub issues and contributing fixes for any Safari-specific bugs that are logged.

## When?

Arguably this is just a matter of updating documentation to reflect the current reality. I intend to wait about a month and then do that in the next releases of jasmine-core and jasmine-browser-runner, unless I hear both strong objections and viable alternatives.

**Comment by triskweline** [comment]
> I intend to wait about a month and then do that in the next releases of jasmine-core and jasmine-browser-runner, unless I hear both strong objections and viable alternatives.

Maybe one alternative: I have a large Jasmine suite that I need to test against various Safari versions. Since I don't have access to a Mac, I use [BrowserStack](https://www.browserstack.com/) to test against a Mac running in the cloud.

Cost: They have an [Open Source program](https://www.browserstack.com/open-source) where you can apply for free access.

Developer experience: BrowserStack is used with a browser. You can select any MacOS and Safari version. Then you get a remote screen within your browser. You can download a binary that lets the remote Safari access your `localhost`.

**Reply by sgravrock** [reply]
Have you been able to set up fully automated CI test runs that use BrowserStack? It seems like it _should_ work: it's supposed to be just another remote Selenium grid. But every time I dig through their docs trying to find out how to set up a proxy like Sauce Connect, I keep coming back to a recommendation to run a wrapper tool that looks like it knows how to make specific supported testing frameworks connect to BrowserStack.

It would be nice to have CI run tests against Safari 18, and I'm starting to wonder if Saucelabs is ever going to get there.

**Reply by triskweline** [reply]
> Have you been able to set up fully automated CI test runs that use BrowserStack? 

[A colleague once did this](https://github.com/makandra/cloud_test), but he used Ruby and Capybara to drive the test suite.

I'm going to look into configuring [Unpoly](https://github.com/unpoly/unpoly)'s CI so it talks to BrowserStack for Safari testing. It currently uses Puppeteer to talk to a local Chrome. I may have to move away from that as Puppeteer only speaks CDP.

> But every time I dig through their docs trying to find out how to set up a proxy like Sauce Connect, I keep coming back to a recommendation to run a wrapper tool that looks like it knows how to make specific supported testing frameworks connect to BrowserStack.

I haven't configured a proxy with BrowserStack before. They have this `BrowserStackLocal` binary that exposes your `localhost` to the remote BrowserStack session.

**Reply by sgravrock** [reply]
> They have this BrowserStackLocal binary that exposes your localhost to the remote BrowserStack session

That's probably the thing I'm looking for. Thanks. I'll give that a try sometime.

**Reply by sgravrock** [reply]
BrowserStackLocal is what I'm looking for, but I can't figure out how to get it to work. The best I've been able to do is get jasmine-browser-runner to fail with "batchReporter is not defined". That's usually a symptom of the browser loading some kind of error page instead of SpecRunner.html. And indeed, when I click on one of my failed test runs in the BrowserStack console, I see that it's loading the Selenium help page. I _think_ that probably happens whether I'm running BrowserStack Local or not, but I can't tell because the other failed runs are all covered by a banner trying to sell me on a paid subscription (minimum $129/month). I already have their OSS sponsorship, but apparently it has major limitations.

Most likely I'm just missing whatever WebDriver capability is needed to tell BrowserStack's grid to connect back to my BrowserStack Local tunnel. But I haven't been able to find any documentation of that. Everything points back to: install our SDK that downloads binaries off the internet and runs them. I'm not going to do that. Even if security wasn't a concern (and it is), I'm not willing to have key parts of Jasmine's own toolchain depend on other people's opaque magic.

All of that could probably be worked around, with enough effort, but putting key diagnostic information behind a paywall is a deal-breaker.

**Reply by sgravrock** [reply]
Summing up the current situation as I see it: although Saucelabs and BrowserStack both provide free services to Jasmine, neither sponsorship provides a useable way to test Jasmine against Safari >= 18 in CI. Unless an alternative is found, it's likely that Safari 17 will be the last Safari version supported by Jasmine.

**Reply by sgravrock** [reply]
> Most likely I'm just missing whatever WebDriver capability is needed to tell BrowserStack's grid to connect back to my BrowserStack Local tunnel. But I haven't been able to find any documentation of that.

Well, past-me must have found it at some point, because I wrote [this](https://github.com/jasmine/jasmine-browser-runner/blob/1a60d4520441321823af95c28c951ebf963c2679/lib/types.js#L217-L229). That's enough to get BrowserStack Local to actually show a connection back from the grid, but the tests still fail to run. I was briefly able to access the results before the paywall kicked in again, but the only debugging information I could find was video of a browser starting up and loading a blank page.It looks like I could enable some things like network traces, console logs, etc via capabilities. But as long as the tools are mostly blocked, it's not worth spending time on.

**Comment by sgravrock** [comment]
> It would be nice to have CI run tests against Safari 18, and I'm starting to wonder if Saucelabs is ever going to get there.

Saucelabs now supports Safari 18, but only on [ARM runners, which are "only available to Enterprise customers with the appropriate Premium plan"](https://docs.saucelabs.com/mobile-apps/virtual-apple-silicon/). So it's safe to say that this is the end of the line for Safari testing on Saucelabs, and we'll need to move to something else like BrowserStack. I haven't been feeling much urgency about this since for now I have a local copy of the only version of Safari that we can't run in CI, but it'll need to be dealt with sooner or later.

**Comment by sgravrock** [comment]
I looked into running browsers directly on CircleCI or GitHub Actions instead of using a remote Selenium grid. Neither is particularly suited to the purpose but GitHub Actions might work as a tool of last resort for running tests on at least the latest version of Safari.

Neither set of images is really intended for browser testing, and neither lets you specify the Safari version directly. With CircleCI you specify the Xcode verison and with GitHub Actions you specify the Mac OS version. At least in GitHub's case the image documentation is sometimes wrong about which Safari version is included. I'm guessing that's because the images are continuously updated.

The more recent CircleCI macos images use a runner class that isn't available to free accounts, making them no more usable than Saucelabs.

GitHub Actions has no such problem. But all except one of its images have Safari 26 installed. The lone image that still has Safari 18 is already in a brownout period and will be removed in early December. So switching to GitHub Actions would buy less than a month before I'd be back to having no overlap between Safari versions we can use in CI and what I have locally.

**Reply by triskweline** [reply]
Thanks for taking the time to look into this and write up your result. Automating tests in Safari is a pain point in many projects for me.

**Reply by sgravrock** [reply]
Thanks, the fact that it's a pain point for you is reassuring in a way.

I also took a look at what other projects are doing, to make sure I wasn't missing a trick. Here's what I found:

* Karma died before this became an issue and its CI looks badly bit-rotted, but it used to use BrowserStack.
* Mocha's cross-browser CI testing gear appears broken and unused. It looks like it used to use Saucelabs, so they'd have the same problem testing newer Safari versions. Policy is that ["Mocha supports the latest major versions of evergreen browsers available when Mocha’s oldest supported Node.js major version was released"](https://mochajs.org/#running-mocha-in-the-browser), which effectively rules out support for any current browsers. I found some comments saying that they turned off browser testing on purpose and they're considering a move to GitHub Actions.
* Vitest uses the macos-latest GitHub Actions runner, so at most they're testing against Safari 26.
* WebdriverIO uses both the macos-13 and macos-latest GitHub Actions runners, probably in an effort to test the widest possible range of Safari versions. I wouldn't be surprised if they also have BrowserStack set up in CI but I can't find it.

Looks like I didn't miss much, and Jasmine wouldn't be out of line to eventually drop support for every Safari version except what's on GitHub Actions. I used to think Jasmine had a problem but now I think Apple has a problem.
